### ISSUE TITLE
Beat singularity-tools up to shape

### ISSUE THREAD

**Issue by SomeoneSerge** [NixOS/nixpkgs#177908]
## Issue description

I intend to start using nixpkgs' `singularity-tools` for hpc applications.
What follows is a list of hindrances and minor annoyances that I've immediately encountered.
The list is mostly for myself: I'm opening the issue to make this visible and maybe motivate people to voice ideas.
Cf. this read on [singularity with Nix](https://discourse.nixos.org/t/nice-read-about-nix-being-used-with-singularity-containers/8973) for more inspiration

- [ ] Get this merged: https://github.com/NixOS/nixpkgs/pull/158486
- [ ] Annoyance: `singularity-tools.buildImage` throws [obscure errors](https://github.com/NixOS/nixpkgs/issues/78743) when `contents = [ ... bash ]`
  - it should instead respond with "$filename already present in the image"
  - it should probably not throw an error if `bash` in `contents` is the same as what the image has been bootstrapped with
- [ ] Annoyance: `mkfs` generates random `UUID`s; make deterministic
- [ ] Annoyance: we can compute `diskSize` from the built `contents` instead of choosing an arbitrary constant
- [ ] Hindrance: failing to pack any cuda-enabled dependencies. The error says: `... Cannot allocate memory`. My `/tmp` is on disk, and I don't seem to be running out of RAM, so this message might be just another version of "not enough space left on (squashfs) device"
- [ ] ...

**Comment by ShamrockLee** [comment]
I would probably have more spare time after June 21.

To speed up the merging of #158486, I would prefer limiting the scope of this PR to:
1.  Support multiple sources and non-vendored source build while
2.  Not to break the previous functionality of `singularity`, `singularity.nix` and `singularity-tools`.

Further improvements can be done in successive PRs.

**Comment by jbedo** [comment]
I agree with limiting the scope of the PR, I'll have time to help in a couple of weeks.

**Comment by ShamrockLee** [comment]
> [ ] Annoyance: we can compute `diskSize` from the built `contents` instead of choosing an arbitrary constant

Is there a way to compute `diskSize` from `contents` at eval time with no IFD?

**Comment by SomeoneSerge** [comment]
> Is there a way to compute diskSize from contents at eval time with no IFD?

I cannot say what's a good way to compute it, but the trivial baseline is a derivation that takes, in `buildInputs`, a `buildEnv` with to-be image's `contents`, and `du`s it. The output times some constant is an upperbound on the `diskSize`

**Comment by ShamrockLee** [comment]
> EDIT: i.e. we wouldn't know `diskSize` at nix eval time, but we'd know it at build time, which appears to be sufficient

Then we would no longer be able to use
```Nix
vmTools.runInLinuxVM (
  runCommand {
    preVM = vmTools.createEmptyImage {
      size = diskSize;
      fullName = "${projectName}-run-disk";
    };
  } ''
    mkfs -t ext3 -b 4096 /dev/${vmTools.hd}
    mount "/dev/${vm.hd}" disk
  ''
)
```

**Comment by dmadisetti** [comment]
For consistency with `dockerTools.buildImage` it would also be nice to change `content` -> `copyToRoot`.

@SomeoneSerge any other HPC pain points?

**Comment by ShamrockLee** [comment]
I don't mind changing the interface of singularity-tools. (That would be a breaking change.)

There's another PR lineing up that builds the image through a Singularity definition (Apptainer recipe) file to make the image more declarative and the build process explainable. It could be a drop-in replacement of the current Singularity-sandbox-based implementation.

I also went on and made a `generator` function that turns a `settings`-like Nix attrset into a definition string. The parser, which does the reverse) is still work in progress.

**Comment by SomeoneSerge** [comment]
_I'm sorry for the long absence, my priorities had shifted somewhat_

@dmadisetti On the high-level I've exactly one pain-point, and that is an unsolved (underinvested) use-case:

- [ ] I want to use singularity to bind-mount `/nix/store` on a cluster that doesn't support user namespaces nor overlayfs, but has a setuid singularity binary
- [ ] I want to ship a pre-built Nix in a singularity image
- [ ] I want to be able to build that image using Nix, e.g. via `singularity-tools.buildImage`

I think I might give this a shot again. The issues I had were:
- [ ] As I said, cluster's singularity installation doesn't come with `--overlay` enabled, so I have to use `--bind`
- [ ] Using `--bind /tmp/blah:/nix/store` hides the container's `/nix/store` -> `singularity run` fails unable to locate the sym-linked `sh` and such.
- [ ] Because `singularity-tools.buildImage` doesn't give user full control over `contents`,  I cannot easily replace the whole thing with static coreutils and a static Nix

Shouldn't be hard to alleviate

**Comment by SomeoneSerge** [comment]
This now suggests another point, that we maybe want a `buildImage` that is extendable and overridable, including the possibility to override the default `contents`. One direction could be `makeOverride`, and I think there's a similar effort being undertaken for `dockerTools`:  https://github.com/NixOS/nixpkgs/pull/208944

Another possibility is the module system with support for `mkMerge`/`mkForce` etc, similar to NixOS. This could also be a viable approach to re-implement the upstream's "definition" files in pure Nix, so as to achieve a declarative interface to `buildImage`.

@ShamrockLee, `settings` approach sounds great, I think this should feel very native in Nixpkgs. Has your work gone into any PRs yet?

**Comment by ShamrockLee** [comment]
> @ShamrockLee, `settings` approach sounds great, I think this should feel very native in Nixpkgs. Has your work gone into any PRs yet?

Not yet, but I already have the implementation integrated the change into my HEP analysis workflow.

It's time to also re-think about the `buildImage` interface IMO.

**Comment by SomeoneSerge** [comment]
@ShamrockLee are you on any of the nixos matrix channels, btw?

**Comment by SomeoneSerge** [comment]
There's a bunch options, including a web client. I think you can just follow either of the links

- [https://matrix.to/#/#dev:nixos.org](https://matrix.to/#/%23dev:nixos.org)
- [https://matrix.to/#/#hpc:nixos.org](https://matrix.to/#/%23hpc:nixos.org)

**Comment by SomeoneSerge** [comment]
By the way, I was meaning to ask, why do we have to `runInLinuxVM`? I remember seeing @jbedo mention this allowed setting setuid flags, but I'm not sure where do we need them. I presume QEMU takes its performance toll

---

P.S. @dmadisetti I guess your post further proves there's a use-case:)

**Comment by ShamrockLee** [comment]
> By the way, I was meaning to ask, why do we have to `runInLinuxVM`?

It is not until last year that the unprivileged image-building workflow started to be implemented in the Apptainer project. The program used to assert `UID == 0` when building the image.

We are closed to the unprivileged image generation with Apptainer. The remaining obstacle is its use of `/var/apptainer/mnt/session` as the container mount point.

See https://github.com/apptainer/apptainer/issues/215

Sylabs's Singularity fork seems to have caught up some progress in unprivileged image build, but it still expects a bunch of top-level directories `/var/singularity/mnt/{container,final,overlay,session,source}`, IIRC.

**Comment by SomeoneSerge** [comment]
> It was not until last year that the unprivileged image-building workflow started to be implemented in the Apptainer project. The program used to assert UID == 0 when building the image.

I see. So, in principle, we could have run everything except `${projectName} build $out ./img` outside QEMU?

**Comment by ShamrockLee** [comment]
> I see. So, in principle, we could have run everything except `${projectName} build $out ./img` outside QEMU?

It's true when it comes to the definition-based build. It won't help much, since it should be trivial in terms of resources to generate the definition file from the definition attrset.

As for the current, Apptainer-sandbox-based `buildImage`, I'm not sure if we could run the `ushare ...` lines for `runAsRoot` outside QEMU.

**Comment by SomeoneSerge** [comment]
> It won't help much

I was rather wondering if we could prepare the file-tree outside qemu and somehow pack the whole batch into an ext3/squashfs image without the `mount`. But then again, I didn't measure, maybe that too is insignificant

**Comment by posch** [comment]
> > It won't help much
> 
> I was rather wondering if we could prepare the file-tree outside qemu and somehow pack the whole batch into an ext3/squashfs image without the `mount`. But then again, I didn't measure, maybe that too is insignificant

I also prefer an approach that doesn't involve creating and running virtual machines. singularity/apptainer can run filesytems in squashfs, and I use this script to create containers:

```
{ pkgs
, contents
, runscript ? "#!/bin/sh\nexec ${pkgs.hello}/bin/hello"
, startscript ? "#!/bin/sh\nexec ${pkgs.hello}/bin/hello"
}:
pkgs.runCommand "make-container" {} ''
  set -o pipefail
  closureInfo=${pkgs.closureInfo { rootPaths=contents ++ [pkgs.bashInteractive]; }}
  mkdir -p $out/r/{bin,etc,dev,proc,sys,usr,.singularity.d/{actions,env,libs}}
  cd $out/r
  cp -na --parents $(cat $closureInfo/store-paths) .
  touch etc/{passwd,group}
  ln -s /bin usr/
  ln -s ${pkgs.bashInteractive}/bin/bash bin/sh
  for p in ${pkgs.lib.concatStringsSep " " contents}; do
    ln -sn $p/bin/* bin/ || true
  done
  echo "${runscript}" >.singularity.d/runscript
  echo "${startscript}" >.singularity.d/startscript
  chmod +x .singularity.d/{runscript,startscript}
  cd $out
  ${pkgs.squashfsTools}/bin/mksquashfs r container.sqfs -no-hardlinks -all-root
  ''
```

**Comment by ShamrockLee** [comment]
FYI: With https://github.com/apptainer/apptainer/pull/1284, Apptainer images can be built as a derivation without a VM.

The code already works (tested with singularity-tools.buildImageFromDef from #224636 specifying `buildImageFlags = [ "--resolv ${pkgs.emptyFile}" "--hosts ${pkgs.emptyFile}" ];`).

The upstream maintainer expects something more general (such as `--no-mount`), so the current change is not likely to get accepted. Nevertheless, it proves that fully-unprivileged Apptainer image build is possible.

**Comment by jbedo** [comment]
> _I'm sorry for the long absence, my priorities had shifted somewhat_
> 
> @dmadisetti On the high-level I've exactly one pain-point, and that is an unsolved (underinvested) use-case:
> 
> * [ ]  I want to use singularity to bind-mount `/nix/store` on a cluster that doesn't support user namespaces nor overlayfs, but has a setuid singularity binary
> * [ ]  I want to ship a pre-built Nix in a singularity image
> * [ ]  I want to be able to build that image using Nix, e.g. via `singularity-tools.buildImage`
> 
> I think I might give this a shot again. The issues I had were:
> 
> * [ ]  As I said, cluster's singularity installation doesn't come with `--overlay` enabled, so I have to use `--bind`
> * [ ]  Using `--bind /tmp/blah:/nix/store` hides the container's `/nix/store` -> `singularity run` fails unable to locate the sym-linked `sh` and such.
> * [ ]  Because `singularity-tools.buildImage` doesn't give user full control over `contents`,  I cannot easily replace the whole thing with static coreutils and a static Nix
> 
> Shouldn't be hard to alleviate

It's a bit hacky but I think this achieves your goals:
```
singularity-tools.buildImage {name = "minimal-nix"; runAsRoot = "${rsync}/bin/rsync -a ${pkgsStatic.nix}/ ./";}
```

**Comment by ShamrockLee** [comment]
> @ShamrockLee are you on any of the nixos matrix channels, btw?

@SomeoneSerge I finally got a Matrix account and join the Nix HPC room, thanks to the Summer of Nix.

**Comment by PhDyellow** [comment]
I managed to get a CUDA-capable container built by adjusting `memSize` along with `diskSize`.

Running it with env vars isn't solved yet.

**Comment by ShamrockLee** [comment]
Let's land #268199 by splitting it into smaller PRs. We could then add the unprivileged Apptainer image build flow as one of its reusable components.

Here's the first one: #332168

**Comment by ShamrockLee** [comment]
> Let's land #268199 by splitting it into smaller PRs. We could then add the unprivileged Apptainer image build flow as one of its reusable components.

#332437 is the second one, containing various fixes and a few deprecations.

**Comment by SomeoneSerge** [comment]
By the way, maybe we should consider dropping support for choosing between apptainer and singularity for building images. For one thing, I suspect we'll have to introduce a separate attribute (like `_apptainer-derandomized` or `_siftool-derandomized`) for a tool patched to leave out all the UUIDs and the timestamps, and it's probably not worth it to maintain patches for both forks...

**Comment by pbsds** [comment]
If the images by one can be ran by the other and are expected to so so going forward then i don't see a problem with that.

**Comment by ShamrockLee** [comment]
> For one thing, I suspect we'll have to introduce a separate attribute (like `_apptainer-derandomized` or `_siftool-derandomized`; #279250) for a tool patched to leave out all the UUIDs and the timestamps, and it's probably not worth it to maintain patches for both forks...

How does patching Apptainer and SingularityCE (the `apptainer` and `singularity` part) make it difficult to choose between Apptainer and SingularityCE for building images (the `singularity-tools` part)? We could define `apptainer` and `singularity` separately if their build flow differs too much while maintaining only one `singularity-tools` for the command-line interface they share in common.

**Comment by SomeoneSerge** [comment]
>      How does patching Apptainer and SingularityCE (the apptainer and singularity part) make it difficult to choose between Apptainer and SingularityCE for building images (the singularity-tools part)? 

It doesn't, it's just that why would we patch them both separately, if we only really need the patches for `singularity-tools`, not for the user-facing singularity?

> If the images by one can be ran by the other and are expected to do so going forward then i don't see a problem with that.

We could even package siftool separately, and that could be enough...

**Comment by ShamrockLee** [comment]
It would be a lot more convenient if the reproducible image build functionality could be implemented upstream.

> We could even package siftool separately, and that could be enough...

I seems to lose track of this. What is siftool?

**Comment by viperML** [comment]
Guix is able to generate to generate `*.sqsh` images directly without QEMU. As far as I know, this is an alternative (older?) format to the `.sif`: https://apptainer.org/docs/user/main/cli/apptainer_shell.html

**Comment by SomeoneSerge** [comment]
> I seems to lose track of this. What is siftool?

Well AFAIU apptainer (singularity? both? not sure, can't remember) uses this go library which comes with a more minimal CLI tool as well:

https://github.com/sylabs/sif/blob/bd743d5fdcfc41372cb45aea3521c206b5afc7af/pkg/siftool/new.go#L15
